# Finding surveillance planes using random forests

**The story:**

- https://www.buzzfeednews.com/article/peteraldhous/spies-in-the-skies
- https://www.buzzfeednews.com/article/peteraldhous/hidden-spy-planes
    
This story, done by Peter Aldhous at Buzzfeed News, involved training a machine learning algorithm to recognize government surveillance planes based on what their flight patterns look like.

**Topics:** Random Forests

**Datasets**

* **feds.csv:** Transponder codes of planes operated by the federal government
* **planes_features.csv:** various features describing each plane's flight patterns
* **train.csv:** a labeled dataset of transponder codes and whether each plane is a surveillance plane or not
    - The `label` column was originally `class`, but I renamed it because pandas freaks out a bit with a column named `class`
    - This was created by Buzzfeed `feds.csv`
* **data dictionary:** You can find the data dictionary published with their analysis [here](https://buzzfeednews.github.io/2016-04-federal-surveillance-planes/analysis.html)
* **a few other files**

## What's the goal?

The FBI and Department of Homeland Security operate many planes that are not directly labeled as belonging to the government. If we can uncover these planes, we have a better idea of the surveillance activities they are undertaking.

## Imports

Also set a large number of maximum columns.

In [23]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from statsmodels.sandbox.regression.predstd import wls_prediction_std
%matplotlib inline

pd.set_option("display.max_columns", 200)
pd.set_option("display.max_colwidth", 200)

# Read in our data

Almost all classification problems start with a set of labeled features. In this case, the features are in one CSV file and the labels are in another. **Read both files in and merge them on `adshex`, the transpoder code.**

In [2]:
features = pd.read_csv('data/planes_features.csv')
features.head()

adshex  duration1  duration2  duration3  duration4  duration5    boxes1  \
0       A   0.120253   0.075949   0.183544   0.335443   0.284810  0.088608   
1  A00000   0.211735   0.155612   0.181122   0.198980   0.252551  0.204082   
2  A00002   0.517241   0.103448   0.103448   0.103448   0.172414  0.862069   
3  A00008   0.125000   0.041667   0.208333   0.166667   0.458333  0.125000   
4  A0001E   0.100000   0.200000   0.200000   0.400000   0.100000  0.100000   

     boxes2    boxes3    boxes4    boxes5    speed1    speed2    speed3  \
0  0.044304  0.069620  0.120253  0.677215  0.021824  0.020550  0.062330   
1  0.183673  0.168367  0.173469  0.267857  0.107348  0.143410  0.208139   
2  0.137931  0.000000  0.000000  0.000000  0.990792  0.000921  0.000000   
3  0.083333  0.125000  0.166667  0.500000  0.187960  0.278952  0.221048   
4  0.000000  0.100000  0.400000  0.400000  0.007937  0.026984  0.084127   

     speed4    speed5  altitude1  altitude2  altitude3  altitude4  altitude5  \
0  0.100713  0.794582   0.042374   0.060971   0.066831   0.106403   0.723421   
1  0.177013  0.364090   0.177318   0.114457   0.129648   0.197694   0.380882   
2  0.000000  0.008287   0.599448   0.400552   0.000000   0.000000   0.000000   
3  0.190257  0.121783   0.014706   0.053309   0.149816   0.279871   0.502298   
4  0.179365  0.701587   0.041270   0.085714   0.039683   0.111111   0.722222   

     steer1    steer2    steer3    steer4    steer5    steer6    steer7  \
0  0.020211  0.048913  0.270550  0.344090  0.097317  0.186651  0.011379   
1  0.034976  0.048127  0.240732  0.356314  0.116116  0.159325  0.012828   
2  0.105893  0.090239  0.174954  0.244015  0.034070  0.202578  0.021179   
3  0.029871  0.044118  0.202665  0.380515  0.094669  0.182904  0.014706   
4  0.019048  0.049206  0.249206  0.326984  0.112698  0.206349  0.012698   

     steer8  flights  squawk_1  observations  type  
0  0.009426      158         0         11776  GRND  
1  0.013628      392         0         52465  TBM7  
2  0.068140       29         0          1086  SHIP  
3  0.020221       24         0          2176  PA46  
4  0.011111       10      1135           630  C56X

In [3]:
labels = pd.read_csv('data/train.csv')
labels.head()

adshex    label
0  A00C4B  surveil
1  A0AB21  surveil
2  A0AE77  surveil
3  A0AE7C  surveil
4  A0C462  surveil

In [4]:
adshex = pd.merge(features, labels, on='adshex', how='left')

### No wait, merge them again!

We have features for about 20,000 planes and labels for about 600 planes. When you merge, the planes you have features for but not labels for will disappear.

We want to keep those in the dataframe so we can play detective with them later, and try to find surveillance planes using the features. When you merge, you should use `how='left'` or `how='right'` to keep unmatched columns from the left (or right) dataframe.

In [5]:
adshex.shape

(19799, 34)

Confirm you have 19,799 rows and 34 columns.

# Cleaning up our data

## Number-izing our labels

Each row is a plane, and it's marked as either a surveillance plane or not. How many do we have in each category?

In [7]:
adshex.label.value_counts()

other      500
surveil     97
Name: label, dtype: int64

How do you feel about that split?

**Prepare this column for machine learning.** What's wrong with it as `"surveil"` and `"other"`? Add a new column that we can use for classification.

In [8]:
adshex['label'] = adshex.label.replace({
    'surveil': 1, 
    'other': 0})

In [9]:
adshex.label.value_counts()


0.0    500
1.0     97
Name: label, dtype: int64

## Categorical variables

Do we have any variables that count as categories? Yes, we do! ...but how many different categories does it have?

* **Tip:** You can use `.unique()` or `.value_counts()` to count unique items, depending on what you're looking for

In [15]:
adshex.type.value_counts()

unknown    2528
C172       1014
SR22        799
BE36        699
C182        693
PC12        630
BE20        465
PA46        438
P28A        415
BE9L        408
C208        395
BE58        387
M20P        335
B350        333
P46T        285
C82S        235
T206        234
C421        225
BE35        222
DA40        214
BE55        204
TBM8        204
B407        202
C340        199
COL4        196
M20T        180
C414        177
AS50        175
EA50        167
PA32        161
           ... 
BEAR          1
BEE3          1
P31           1
ONEX          1
BN2T          1
BOLT          1
BRAV          1
BREZ          1
BT36          1
C10T          1
MSQ2          1
MS76          1
C162          1
MIMU          1
MG17          1
MC90          1
M5            1
M4            1
M20J          1
M20F          1
C205          1
M20           1
M16           1
LJ55          1
C25           1
LJ24          1
L90           1
KIS4          1
J2            1
208           1
Name: type, Length: 455,

Most of those types of plane only have one appearance, which means they wouldn't be very helpful identifiers in the final analysis. For example, if I only see one GLF5 and it's a surveillance plane, does that mean the next one I see is probably a surveillance plane? With such a small sample size, I have no idea!

We have a few options

1. Create a very large set of dummy variables out of all 133 types of planes
2. Create `0`/`1` columns for common plane types and ignore the less common ones -  C182, T206, SR22
3. Interview someone who knows something about planes and put these into a few broader categories
4. Keep them as one column, just turn them into numbers - it doesn't make sense in terms of order, but if one or two plane types are very indicative of a surveillance plane the forest might pick it up

Oddly enough, **the last one is a common approach.** Let's use it!

If you want to convert a list of categories into numbers, an easy way is to use the `Categorical` data type.

In [14]:
adshex.type = adshex.type.astype('category')
adshex.type.head()

0    GRND
1    TBM7
2    SHIP
3    PA46
4    C56X
Name: type, dtype: category
Categories (455, object): [208, A109, A119, A139, ..., WW24, XL2, ZZZZ, unknown]

It looks like a normal bunch of strings, but pandas is secretly using a number for each one! You can find the number with `.cat.codes`.

**Use `df.type.cat.codes` to make a new columns called `type_code`.** 

In [17]:
adshex['type_code'] = adshex.type.cat.codes

We'll use `type_code` for machine learning since sklearn needs a number, and `type` for reading since we like text.

# Building our classifier

When we're about to classify, we usually just drop our target column to build our inputs and outputs:

```python
X = train_df.drop(column='column_you_are_predicting')
y = train_df.column_you_are_predicting
```

This time is a little different. First, we have unlabeled data in there! Use `.dropna()` to filter your training data so we only have labeled data.

Confirm `train_df` has 597 rows and 35 columns.

In [19]:
adshex.head(3)

adshex  duration1  duration2  duration3  duration4  duration5    boxes1  \
0       A   0.120253   0.075949   0.183544   0.335443   0.284810  0.088608   
1  A00000   0.211735   0.155612   0.181122   0.198980   0.252551  0.204082   
2  A00002   0.517241   0.103448   0.103448   0.103448   0.172414  0.862069   

     boxes2    boxes3    boxes4    boxes5    speed1    speed2    speed3  \
0  0.044304  0.069620  0.120253  0.677215  0.021824  0.020550  0.062330   
1  0.183673  0.168367  0.173469  0.267857  0.107348  0.143410  0.208139   
2  0.137931  0.000000  0.000000  0.000000  0.990792  0.000921  0.000000   

     speed4    speed5  altitude1  altitude2  altitude3  altitude4  altitude5  \
0  0.100713  0.794582   0.042374   0.060971   0.066831   0.106403   0.723421   
1  0.177013  0.364090   0.177318   0.114457   0.129648   0.197694   0.380882   
2  0.000000  0.008287   0.599448   0.400552   0.000000   0.000000   0.000000   

     steer1    steer2    steer3    steer4    steer5    steer6    steer7  \
0  0.020211  0.048913  0.270550  0.344090  0.097317  0.186651  0.011379   
1  0.034976  0.048127  0.240732  0.356314  0.116116  0.159325  0.012828   
2  0.105893  0.090239  0.174954  0.244015  0.034070  0.202578  0.021179   

     steer8  flights  squawk_1  observations  type  label  type_code  
0  0.009426      158         0         11776  GRND    NaN        248  
1  0.013628      392         0         52465  TBM7    NaN        431  
2  0.068140       29         0          1086  SHIP    0.0        399

In [21]:
train_df = adshex.dropna()

We also have a few extra columns that we aren't using for classification (like the text version of the type column and the transponder code). It's fine to drop multiple columns here that you aren't using, just a little bit messier. You also have to make sure you're dropping all the right ones.

Do a `.head()` to double-check all of the columns you need to drop when creating your `X`.

In [22]:
train_df = train_df.drop(['adshex', 'type'], axis=1)

### Create your `X` and `y`.

When you do `train_df.drop`, you'll want to remove more than just your `0`/`1` surveillance label. What other columns do you not want to use as input? Maybe some categories you converted into codes?

In [24]:
X = train_df.drop(columns='label')
y = train_df.label

Triple-check that `X` is a list of numeric features and and `y` is a numeric label.

### Split into test and train datasets

We could be nice and lazy and use all our data for training, but it just isn't right! Taking a test using the exact same questions you studied is just cheating. Split your data into test and train.

* **Tip:** Don't do this manually! There's a method for it in sklearn

In [25]:
from sklearn import tree
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)

In [28]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(447, 32)
(150, 32)
(447,)
(150,)


# Classify using a logistic classifier

## Train your classifier

Build a `LogisticRegression` and fit it to your data, making sure you're training using only `X_train` and `y_train`.

* **Tip:** You'll want to give `LogisticRegression` an extra argument of `max_iter=4000` - it means "work a little harder than you expect," because otherwise it won't find an answer (by default it only has a `max_iter` of 100)

In [30]:
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(C=1e9, solver='lbfgs', max_iter=4000)

clf.fit(X, y)

feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients)
}).sort_values(by='odds ratio', ascending=False)

feature  coefficient (log odds ratio)  odds ratio
10        speed1                      0.625828    1.869794
21        steer2                      0.520862    1.683478
17     altitude3                      0.436915    1.547924
5         boxes1                      0.386663    1.472061
20        steer1                      0.329899    1.390827
6         boxes2                      0.314007    1.368899
0      duration1                      0.111638    1.118108
27        steer8                      0.006517    1.006539
29      squawk_1                      0.000702    1.000702
31     type_code                      0.000409    1.000409
30  observations                      0.000010    1.000010
28       flights                     -0.001029    0.998971
26        steer7                     -0.007373    0.992655
11        speed2                     -0.123287    0.884010
1      duration2                     -0.172444    0.841606
18     altitude4                     -0.222903    0.800193
7         boxes3                     -0.347349    0.706559
22        steer3                     -0.394305    0.674148
16     altitude2                     -0.423640    0.654660
4      duration5                     -0.456116    0.633740
24        steer5                     -0.469731    0.625170
25        steer6                     -0.503407    0.604468
12        speed3                     -0.545365    0.579630
2      duration3                     -0.628887    0.533185
15     altitude1                     -0.682416    0.505394
8         boxes4                     -0.804288    0.447406
3      duration4                     -0.824638    0.438394
13        speed4                     -0.880102    0.414740
14        speed5                     -1.047521    0.350806
19     altitude5                     -1.078403    0.340138
23        steer4                     -1.501197    0.222863
9         boxes5                     -1.519481    0.218826

## Examine the coefficients

What does it mean? What features is the classifier using? Do you care about the odds ratio? **What is even the point of this `LogisticRegression` thing?**

In [31]:
import numpy as np

feature_names = X.columns
coefficients = clf.coef_[0]

pd.DataFrame({
    'feature': feature_names,
    'coefficient (log odds ratio)': coefficients,
    'odds ratio': np.exp(coefficients)
}).sort_values(by='odds ratio', ascending=False)

feature  coefficient (log odds ratio)  odds ratio
10        speed1                      0.625828    1.869794
21        steer2                      0.520862    1.683478
17     altitude3                      0.436915    1.547924
5         boxes1                      0.386663    1.472061
20        steer1                      0.329899    1.390827
6         boxes2                      0.314007    1.368899
0      duration1                      0.111638    1.118108
27        steer8                      0.006517    1.006539
29      squawk_1                      0.000702    1.000702
31     type_code                      0.000409    1.000409
30  observations                      0.000010    1.000010
28       flights                     -0.001029    0.998971
26        steer7                     -0.007373    0.992655
11        speed2                     -0.123287    0.884010
1      duration2                     -0.172444    0.841606
18     altitude4                     -0.222903    0.800193
7         boxes3                     -0.347349    0.706559
22        steer3                     -0.394305    0.674148
16     altitude2                     -0.423640    0.654660
4      duration5                     -0.456116    0.633740
24        steer5                     -0.469731    0.625170
25        steer6                     -0.503407    0.604468
12        speed3                     -0.545365    0.579630
2      duration3                     -0.628887    0.533185
15     altitude1                     -0.682416    0.505394
8         boxes4                     -0.804288    0.447406
3      duration4                     -0.824638    0.438394
13        speed4                     -0.880102    0.414740
14        speed5                     -1.047521    0.350806
19     altitude5                     -1.078403    0.340138
23        steer4                     -1.501197    0.222863
9         boxes5                     -1.519481    0.218826

## How well does our classifier perform?

Let's take a look at the confusion matrix to see how well this classifier finds surveillance planes.

```python
from sklearn.metrics import confusion_matrix

y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not surveil', 'surveil'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)
```

Notice we're using `y_test` and `X_test`, not the full dataset.

In [32]:

from sklearn.metrics import confusion_matrix

y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not surveil', 'surveil'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)

Predicted not surveil  Predicted surveil
Is not surveil                    124                  3
Is surveil                         10                 13

Why do we use `y_test` and `X_test` instead of the full dataset?

In [ ]:
#to test our model. how well it was trained by the training data..

# Classify using a decision tree

Now we'll use a decision tree. This is how you make one:

```python
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier()
```

But it's up to you to teach it what spy planes look like using your training data.

If we use `max_depth=` to limit the depth of the tree, it will help us visualize it. For example, `max_depth=5` will only allow the tree to make five decisions.

Make a decision tree and fit it to your data. Use a `max_depth=` of something between 2 to 5.

In [33]:
from sklearn.tree import DecisionTreeClassifier

clf = tree.DecisionTreeClassifier(max_depth=5)
clf = clf.fit(X_train, y_train)

## What are the important features?

This code is slightly different than feature importance for logistic regression. It looks like this:

```python
feature_names = X.columns
importances = clf.feature_importances_

pd.DataFrame({
    'feature': feature_names,
    'feature importance': importances,
}).sort_values(by='feature importance', ascending=False)
```

In [34]:
feature_names = X.columns
importances = clf.feature_importances_

pd.DataFrame({
    'feature': feature_names,
    'feature importance': importances,
}).sort_values(by='feature importance', ascending=False)

feature  feature importance
21        steer2            0.730182
29      squawk_1            0.157719
3      duration4            0.046479
19     altitude5            0.033579
0      duration1            0.011460
18     altitude4            0.008448
28       flights            0.007965
11        speed2            0.004167
27        steer8            0.000000
26        steer7            0.000000
17     altitude3            0.000000
24        steer5            0.000000
23        steer4            0.000000
22        steer3            0.000000
20        steer1            0.000000
30  observations            0.000000
25        steer6            0.000000
16     altitude2            0.000000
1      duration2            0.000000
15     altitude1            0.000000
14        speed5            0.000000
13        speed4            0.000000
12        speed3            0.000000
10        speed1            0.000000
9         boxes5            0.000000
8         boxes4            0.000000
7         boxes3            0.000000
6         boxes2            0.000000
5         boxes1            0.000000
4      duration5            0.000000
2      duration3            0.000000
31     type_code            0.000000

### Understanding the output

**Why is the feature importance difference than for logistic regression?**

Also, if you don't specify a `max_depth`, that's a LOT of zeroes! It doesn't even use most of the features! **Why not?**

In [ ]:
#maybe bc the are not relevant enough for the specified depth to make a difference..

## How well does the tree perform?

Display another confusion matrix with your new classifier.

In [35]:
y_true = y_test
y_pred = clf.predict(X_test)
matrix = confusion_matrix(y_true, y_pred)

label_names = pd.Series(['not surveil', 'surveil'])
pd.DataFrame(matrix,
     columns='Predicted ' + label_names,
     index='Is ' + label_names)


Predicted not surveil  Predicted surveil
Is not surveil                    123                  4
Is surveil                          5                 18

## Visualize the tree

You can use this code to visualize the tree. You might need to `brew install graphviz` and `pip install graphviz`.

```python
from sklearn import tree
import graphviz

label_names = ['not surveillance', 'surveillance']
feature_names = X.columns

dot_data = tree.export_graphviz(clf,
                    feature_names=feature_names,  
                    filled=True,
                    class_names=label_names)  
graph = graphviz.Source(dot_data)  
graph
```

* **Tip:** You'll probably need to scroll sideways a bit

In [ ]:
#cannot brew install....

In [41]:
from sklearn import tree
import graphviz

label_names = ['not surveillance', 'surveillance']
feature_names = X.columns

dot_data = tree.export_graphviz(clf,
                    feature_names=feature_names,  
                    filled=True,
                    class_names=label_names)  
graph = graphviz.Source(dot_data)  
graph

ExecutableNotFound: failed to execute ['dot', '-Tsvg'], make sure the Graphviz executables are on your systems' PATH

# One more classifier: Random forest

## Build and train your classifier

We can build a random forest classifier like this:

```python
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
```

But you're in charge of fitting it to your training data!

* **Tip:** You can also set `max_depth` here, but you won't be able to visualize the result.
* **Tip:** Increase `n_estimators` to 100 to make a better classifier.

In [42]:

from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators = 100, max_depth=5)
model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=5, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

## What are the important features?

In [43]:
feature_names = X.columns
importances = clf.feature_importances_

pd.DataFrame({
    'feature': feature_names,
    'feature importance': importances,
}).sort_values(by='feature importance', ascending=False)

feature  feature importance
21        steer2            0.730182
29      squawk_1            0.157719
3      duration4            0.046479
19     altitude5            0.033579
0      duration1            0.011460
18     altitude4            0.008448
28       flights            0.007965
11        speed2            0.004167
27        steer8            0.000000
26        steer7            0.000000
17     altitude3            0.000000
24        steer5            0.000000
23        steer4            0.000000
22        steer3            0.000000
20        steer1            0.000000
30  observations            0.000000
25        steer6            0.000000
16     altitude2            0.000000
1      duration2            0.000000
15     altitude1            0.000000
14        speed5            0.000000
13        speed4            0.000000
12        speed3            0.000000
10        speed1            0.000000
9         boxes5            0.000000
8         boxes4            0.000000
7         boxes3            0.000000
6         boxes2            0.000000
5         boxes1            0.000000
4      duration5            0.000000
2      duration3            0.000000
31     type_code            0.000000

### Understanding the output

What is a random forest, and **why is the feature importance difference than for the decision tree?** Isn't a random forest just like a decision tree or something?

## How well does it perform?

In [44]:
from sklearn import metrics

y_pred=model.predict(X_test)

print('Accuracy:', metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9666666666666667


### How confident do you feel in the model?

In [ ]:
#thats pretty good.. but maybe overfitting though?

# Actually finding spy planes

Now let's try ot actually find our spy planes

## Retrain our model

When we did test/train split, we trained our model with only a subset of our data, so we could test with the rest. Now that we're working in the "real world" we want to re-train it using not just `_train` and `_test` data, but instead **everything we have labels for.**

In [45]:
clf = RandomForestClassifier(n_estimators = 100, max_depth=5)
clf = clf.fit(X, y)

## Filter for planes we want to predict

We have a dataframe of features that includes three types of planes:

* Those that are labeled as surveillance planes
* Those that are labeled as not surveillance
* Those that aren't labeled

Which do we want to predictions for? **Filter a new dataframe that's just those.**

* **Tip:** Scroll up to see where you created your `train_df`, it's the opposite!

In [46]:
new_df = adshex[adshex.label.isnull()]

How many planes do you have in that list? **Confirm it's about 19,200.**

In [47]:
new_df.shape

(19202, 35)

## Predicting 

Build your `X` - remember you need to drop a few columns - and use that to make a prediction for each plane.

**Assign the prediction into the `predicted` column**.

* **Tip:** Scroll up to see where you created your features for training, it's similar
* **Tip:** pandas will yell at us about setting values on copies of a slice but it's fine

In [48]:

X = new_df.drop(columns=['label', 'adshex', 'type'], axis=1)

## How many planes did it predict to be surveillance planes?

It should be roughly around 70-80 planes.

In [50]:
X['predicted'] = clf.predict(X)


In [51]:
X[X['predicted'] == 1].shape
#seems ok...

(86, 33)

## But.. what about those other ones? The ones that are just below the threshold?

The cutoff for a prediction of `1` is 50%, but since we have a lot of time we're interested in investigating the top 150. To get the probability for each row, you will use `clf.predict_proba` instead of `clf.predict`. Also, to get the predicted probability for the `1` category, you'll need to add `[:,1]` to the end of the

```python
clf.predict_proba(***your features***)[:,1]
```

**Create a new column called `predicted_prob` that is the chance that the plane is a surveillance plane.**

* **Tip:** You dropped three columns when using `clf.predict`, but if you drop the same three you'll get an error now. There's now an extra column that you'll need to drop! What is it?

In [52]:
X = X.drop(columns=['predicted'], axis=1)

In [53]:

X['predicted_prob'] = clf.predict_proba(X)[:,1]

In [54]:
X.predicted_prob.value_counts(ascending=False)

0.003385    224
0.003469     97
0.007804     58
0.003266     51
0.003227     50
0.004248     38
0.007685     37
0.004147     37
0.002861     35
0.003124     33
0.003648     29
0.006058     23
0.003311     22
0.006416     22
0.004323     19
0.003926     17
0.004640     16
0.003530     16
0.017804     16
0.010060     16
0.006500     15
0.007693     15
0.003490     15
0.003005     12
0.004191     12
0.003561     10
0.004180     10
0.003679     10
0.005225     10
0.003208     10
           ... 
0.083317      1
0.014388      1
0.015262      1
0.173875      1
0.023477      1
0.023962      1
0.010364      1
0.082981      1
0.008837      1
0.091736      1
0.030463      1
0.051630      1
0.039964      1
0.032338      1
0.062138      1
0.005824      1
0.031060      1
0.009243      1
0.015401      1
0.042922      1
0.035923      1
0.007417      1
0.049377      1
0.102555      1
0.009635      1
0.027214      1
0.617494      1
0.033211      1
0.182781      1
0.008815      1
Name: predicted_prob, Le

### Get the top 200 predictions

Take a look at what the probabilities look like, showing the top 200 planes that are **most likely to be surveillance planes.**

Then save them to a file for later research.

In [55]:
df = X.sort_values(by='predicted_prob', ascending=False).head(200)

In [56]:
df.to_csv('spy_planes', index=False)
df = pd.read_csv('spy_planes')
df.head()

duration1  duration2  duration3  duration4  duration5    boxes1    boxes2  \
0   0.166667   0.166667   0.166667   0.083333   0.416667  0.250000  0.583333   
1   0.416667   0.125000   0.083333   0.041667   0.333333  0.458333  0.458333   
2   0.615385   0.076923   0.153846   0.000000   0.153846  0.538462  0.307692   
3   0.333333   0.196078   0.137255   0.078431   0.254902  0.549020  0.313725   
4   0.121212   0.141414   0.070707   0.070707   0.595960  0.212121  0.515152   

     boxes3    boxes4  boxes5    speed1    speed2    speed3    speed4  \
0  0.166667  0.000000     0.0  0.866572  0.071664  0.035361  0.020745   
1  0.083333  0.000000     0.0  0.562937  0.226224  0.138811  0.056294   
2  0.076923  0.076923     0.0  0.524156  0.331568  0.103243  0.039709   
3  0.137255  0.000000     0.0  0.817386  0.122310  0.051307  0.005714   
4  0.242424  0.030303     0.0  0.271168  0.494554  0.212671  0.016859   

     speed5  altitude1  altitude2  altitude3  altitude4  altitude5    steer1  \
0  0.005658   0.053748   0.123055   0.665724   0.157473   0.000000  0.151344   
1  0.015734   0.000000   0.009091   0.039860   0.866434   0.084615  0.144406   
2  0.001324   0.053607   0.080079   0.756453   0.105228   0.004633  0.455328   
3  0.003283   0.077204   0.297143   0.622492   0.003161   0.000000  0.207295   
4  0.004747   0.018678   0.065840   0.345793   0.568557   0.001131  0.166840   

     steer2    steer3    steer4    steer5    steer6    steer7    steer8  \
0  0.176803  0.181047  0.300802  0.019331  0.085809  0.010372  0.028289   
1  0.226923  0.222378  0.268182  0.013986  0.062587  0.004196  0.017483   
2  0.173395  0.138981  0.052945  0.021840  0.049636  0.001985  0.008604   
3  0.219210  0.251185  0.125471  0.028571  0.075137  0.009848  0.019574   
4  0.315047  0.301537  0.096653  0.015661  0.047095  0.004015  0.009250   

   flights  squawk_1  observations  type_code  predicted_prob  
0       12      4415          2121        454        0.944832  
1       24      5310          2860        322        0.944096  
2       13      5240          1511         98        0.929469  
3       51      5306          8225        417        0.925259  
4       99       230         45079        417        0.922517

### Question 1

What kind of machine learning are we doing here, and why are we doing it?

### Question 2

What are a few different ways you can deal with categorical data? Think about how we dealt with race in the reveal regression compared to how we dealt with type in this dataset.

### Question 3

Every time we ran a machine learning algorithm on our dataset, we looked at feature importance.

* When might it be important to explain what our model found important?
* When might it not be important?

### Question 4

Using words and not column names, describe what the machine learning algorithm found to be important when identifying surveillance planes.

### Question 5

Why did we use test/train split when it would have been more effective to give our model all of the data from the start?

### Question 6

Why did we use a random forest instead of a decision tree or logistic regression? Was there something about the data?

### Question 7

Why did we use probability instead of just looking for planes with a predicted value of 1? It seems like we should have just trusted the algorithm, right?

### Question 8

What if our random forest or input dataset were flawed? What would be the repercussions?

### Question 9

The government could claim that we're threatening national security by publishing this paper as well as publishing this code - now anyone could look for planes that are surveilling them. What do you think?

### Question 10

We're using data from the past, but you can get real-time flight data from many services. Can you think of any uses for this algorithm using real-time instead of historical data?

### Question 11

This isn't a question, but if you look at `candidates.csv` and `candidates-annotates.csv` you can see how Buzzfeed did their research after finding a list of suspicious planes.